### PR TITLE
Implementation of visualisation tools

### PR DESCRIPTION
**Pull request by wingedRuslan**
- [x] I'm ready to merge

* **What's the context for this pull request?** 
_(this is a good place to reference any issues that this PR addresses)_

Visualisation tools  
References issues #116 #115 and #119 #118

* **What's new?**

I've implemented visualisation tools for the following purposes:

1) `plot_rich_club` - for plotting the rich club values per degree along with the random rich club values created from a random network with a preserved degree distribution.

**Result was before comments**
Rich_club_values

**Result now after comments**
Rich_club_values

2) `plot_network_measures` - plotting network measures values along with the random network values values created from a random network

**Result**
Barrr

3) `plot_degree_dist` - plotting the degree distribution along with the degree distribution of an Erdos Renyi random graph

**Result was before comments**
3333Plot

**Result now after comments (no gap)**

degree_dist

Kirstie's code was really helpful but mostly I've designed these functions by myself, so there are might be some features(changes) you would like to have and on the contrary not to have.

Basically, I tried to follow the Kirstie's code (like to have the same parameters to the functions), but the main implementation was either hard to reproduce or had errors while running. That's why I relied on my intuition and created functions from my perspective.

Feel free to point out me any wrong decisions I took while implementing visualization tools.

I have also tried to have the same outputs illustrated in issue #1.

The way how to use created functionalities I showed in the **visualisations_tutorial**. 

* **What should a reviewer feedback on?**

The design of the functions and the produced figures

* **Does anything need to be updated after merge?** 
_(e.g the wiki or the WhitakerLab website)_

Documentation is updated to include a new module - visualisations.py

### CODE REVIEW

**Comment by KirstieJane** [comment]
This is SO COOL @wingedRuslan! Way to go!

I'll add some comments now 😸

File: scona/visualisations.py [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+    # plot bar chart for network measures
+    rects1 = ax.bar(r1, sorted_net_values, color=color[0], width=barWidth,
+                    edgecolor='white', label='Network Measures')
+
+    # plot bar chart for random network measures
+    rects2 = ax.bar(r2, sorted_random_net_values, color=color[1], width=barWidth,
+                    edgecolor='white', label='Random Network Measures')
+
+    # autolabel each bar column with the value
+    if labelBar:
```
Comment (KirstieJane) [inline scona/visualisations.py]: Very cool ✨

File: scona/visualisations.py [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+    ax.axhline(0, linewidth=0.7, color='black')
+
+    # set the number of bins to 5
+    ax.locator_params(axis='y', nbins=5)
+
+    # set labels for y axix
+    ax.set_ylabel("Network Values")
+
+    # create a legend
+    ax.legend(fontsize="xx-small")
```
Comment (KirstieJane) [inline scona/visualisations.py]: Option, as above.

File: scona/visualisations.py [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+    ax.set_xticklabels(barsLabels)
+
+    # make a line at y=0
+    ax.axhline(0, linewidth=0.7, color='black')
+
+    # set the number of bins to 5
+    ax.locator_params(axis='y', nbins=5)
+
+    # set labels for y axix
+    ax.set_ylabel("Network Values")
```
Comment (KirstieJane) [inline scona/visualisations.py]:
```suggestion
    ax.set_ylabel("Global network measures")
```

File: scona/visualisations.py [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+
+    # plot distribution of graph's degrees
+    ax = sns.distplot(degrees, color=color[0])
+
+    # plot a Erdos Renyi graph density estimate
+    if binomial_graph:
+        ax = sns.kdeplot(degrees_ER, color=color[1])
+
+    # fix the x axis limits
+    ax.set_xlim((0, max(degrees)))
```
Comment (KirstieJane) [inline scona/visualisations.py]: It doesn't look in the figure in the issue like this is starting at 0.... it looks like there's a gap between 0 and the first bin. Maybe set the number of bins rather than letting seaborn decide (https://seaborn.pydata.org/generated/seaborn.distplot.html)?

File: scona/visualisations.py [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+    """
+
+    # get the degrees
+    degree = list(rc_coef.keys())
+
+    # get the rich club coefficient values
+    rc = list(rc_coef.values())
+
+    # get the random rich club coefficient values
+    rc_rand = list(rc_coef_rand.values())
```
[the remark this answers is not in the document]
Comment (Islast) [reply scona/visualisations.py]: So to get these rich club values for your random graphs you would call `B.report_rich_club()` where `B` is your [GraphBundle (see docs)](https://whitakerlab.github.io/scona/scona.html?highlight=graphbundle#scona.classes.GraphBundle) containing all your random graphs.

I think ideally we should be passing a GraphBundle directly to the plotting function, since this is the scona class for handling inter network comparisons, unfortunately the documentation on this subject is a little sparse and I haven't had a chance to talk Ruslan through it yet

File: scona/visualisations.py [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+
+    # set the number of bins to 4
+    ax.locator_params(nbins=4)
+
+    # set the x and y axis labels
+    ax.set_xlabel("Degree")
+    ax.set_ylabel("Rich Club")
+
+    # create a legend
+    ax.legend(fontsize="x-small")
```
[the remark this answers is not in the document]
Comment (wingedRuslan) [reply scona/visualisations.py]: Sure, easily

File: scona/visualisations.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (wingedRuslan) [reply scona/visualisations.py]: @KirstieJane So we need to plot rich club coefficient values obtained not only from one random graph, but rich club coefficient values with error bars from a bunch of random graphs, right? 
(at this moment, we are plotting real rich club values and random rich club values from 1 random graph to compare with)

File: scona/visualisations.py [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+
+    # adjust subplot params so that the subplot fits in to the figure area
+    plt.tight_layout()
+
+    # display the figure
+    plt.show()
+
+    # save the figure if the location-to-save is provided
+    if figure_name:
+        fig.savefig(figure_name, bbox_inches=0, dpi=100)
```
[the remark this answers is not in the document]
Comment (wingedRuslan) [reply scona/visualisations.py]: Ohhh, did not check that, I will modify code to handle this case

File: scona/visualisations.py [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+    figure_name : str, optional
+        path to the file to store the created figure in (e.g. "/home/Desktop/name")
+        or to store in the current directory include just a name ("fig_name");
+    color : list of 2 strings, optional
+        where the 1st string is a color for rich club values and 2nd - for random
+        rich club values. You can specify the color using an html hex string
+        (e.g. color =["#06209c","#c1b8b1"]) or you can pass an (r, g, b) tuple,
+        where each of r, g, b are in the range [0,1]. Finally, legal html names
+        for colors, like "red", "black" and so on are supported.
+    labelBar : bool, optional
```
[the remark this answers is not in the document]
Comment (wingedRuslan) [reply scona/visualisations.py]: sure, I should be consistent in naming

File: scona/visualisations.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (wingedRuslan) [reply scona/visualisations.py]: @Islast @KirstieJane What do you think about passing a **parameter**  num_rand_measures=Integer` to the function and based on the value (0 or 3 or 10) we create the corresponding amount of random graphs and then random measures?
Like this:
```
for i in range(num_rand_measures):
    R = random_graph(H)
    
    # Calculate the rich club coefficient
    rc_rand_dict = rich_club(R)

    # And save the values to the numpy array
    rc_rand[:, i] = list(rc_rand_dict.values())
```
But in my previous implementation, I took the mean of network measures and passed *real_network_measures* and *random_network_measures* to the function to plot. 
But I should plot all the random network measures, right?

File: scona/visualisations.py [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+    """
+    This is a visualisation tool for plotting network measures values
+    along with the random network values values created from a random network.
+
+    Parameters
+    ----------
+    network_measures : dict
+        real network measures
+        Note: the dict could be obtained from calculate_global_measures()
+    rand_network_measures : :dict
```
[the remark this answers is not in the document]
Comment (wingedRuslan) [reply scona/visualisations.py]: Just one. 
Yeah, I did not know that we need a bunch of random values. I've created functions to accept only 1 random values.
> This is the same issue as above (line 49)

Right now `plot_network_measures(network_measures, rand_network_measures..)` accepts
```
network_measures : dict
        real network measures
rand_network_measures : dict (from 1 random graph)
        random network measure values
```

Should I redesign so that the function accepts **rand_network_measures** which is a list of some random measures? 
Like:
```
{'assortativity': x1, x2, ... xn 
 'average_clustering': x1, x2, ... xn 
 'average_shortest_path_length': x1, x2, ... xn 
 'efficiency': x1, x2, ... xn 
 'modularity': x1, x2, ... xn }
```
And then plot with error bars, right?

File: scona/visualisations.py (hunk 4 above)
Comment (wingedRuslan) [reply scona/visualisations.py]: would you like to get rid of a small gap between 0 and the 1st column?

File: scona/visualisations.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Islast) [reply scona/visualisations.py]: Having an option to generate a bunch of random graphs on the fly is not a bad idea. 
The ideal way to generate new random graphs is to initiate a `GraphBundle` object.
```python
brain_bundle = scn.GraphBundle([H], ['trueNetwork'])
brain_bundle.create_random_graphs('trueNetwork', num_random_graphs)
```
You can get a dataframe of rich club values from this by calling `brain_bundle.report_rich_club()`.  When you calculate error bars/mean etc from this dataframe remember to mask out the original graph `'trueNetwork'`.

File: scona/visualisations.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Islast) [reply scona/visualisations.py]: I think it would be ideal to accept the dataframe or dict returned by [`brain_bundle.report_global_measures()`](https://whitakerlab.github.io/scona/scona.html#scona.classes.GraphBundle.report_global_measures) if you can figure that out

File: scona/visualisations.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (wingedRuslan) [reply scona/visualisations.py]: @Islast 
I understood the scona way of dealing with across network comparisons by using GraphBundle.
I have redesigned the plotting functions to use `GraphBundle` object as a parameter. 
! But I have an assumption that the when we calculate the *rich_club_values* (`brain_bundle.report_rich_club()`) the *1st column will be our real network*, all others - random networks. Is this always true? (I followed the process of creating random graphs and reporting rich club, but still, I am not 100% sure that real_network will **always** be the 1st column while **reporting rich club**.

Otherwise, I also have to pass the *name of the Real Network* to plotting function.

Right now the signature is `plot_rich_club(brain_bundle,...)`. If I am mistaken with my a-little-tested assumption, the signature would be `plot_rich_club(brain_bundle, name_realGraph)``

File: scona/visualisations.py (hunk 4 above)
Comment (wingedRuslan) [reply scona/visualisations.py]: I've removed that gap

File: scona/visualisations.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Islast) [reply scona/visualisations.py]: My personal preference in these matters is to be super explicit about everything, so I would pass the name of the real network
If you prefer you could have default set to first column with an option to pass a name - as long as the docstrings are **super clear** about what is going on

[the remark this answers is not in the document]
Comment (wingedRuslan) [reply scona/visualisations.py]: Okay, I will make changes to accept *real_network_name* as a parameter to function

**Comment by KirstieJane** [comment]
Hey @wingedRuslan! I've had a great evening playing with your tutorial!! Super fun 💖 

I've had a big edit of the tutorial - see https://github.com/wingedRuslan/scona/pull/2 for the changes.

There are only small changes in the other parts of the code.

The `plot_network_measures` code _works_ (beautifully) but I find it really hard to read. Can you have a go at adding a bunch more documentation to it? I think I would create another function that builds the pandas data frames. That's the part that I'm really not following well when I read the code.

**Comment by wingedRuslan** [comment]
Thanks for the review @KirstieJane! 

Do you mean to make docstring more detailed and well-explained or/and to make code to be read easily?
>The plot_network_measures code works (beautifully) but I find it really hard to read. Can you have a go at adding a bunch more documentation to it? 

Nice idea! I will create one in helpers.py
>I think I would create another function that builds the pandas data frames.

Yeah, I agree about creating another tutorial for brain network visualisation and I will do that. I really wanted to make one today, but unexpectedly spent a lot of time on colormaps, sooorry :(

**Comment by KirstieJane** [comment]
I’m not super sure why the function needs to be in helpers.py - I’m talking about building the specific data frame for plotting the nodal measures..... I think I’d keep it in visualisations.py because that’s the only place it will be called?? Happy to be convinced otherwise though.

On the documentation front, I mean both: more info in the doc string about what is happening with the function AND making the code more readable 😬

No problem on the timings!! You’ve done tonnes of great work here!

**Comment by KirstieJane** [comment]
OH - and one last thing. Make sure to run the code through flake8 linting is you can. We have quite a lot of long lines at the moment ✏️

**Comment by wingedRuslan** [comment]
Yes, right! it makes sense :)
> I think I’d keep it in visualisations.py because that’s the only place it will be called??

My motivation to add a newly created function ( that builds the specific data frame for plotting the global measures) to the helpers.py because it will make files, modules (and work?) more organized.
I mean that in `visualisations.py` we have only functions to make plots and nothing else, in `helpers.py` -  any other functions needed for plotting ("helpers" functions that help to make plots and not only plots). But you are right, this function that builds a specific dataframe for plotting global measures will be used only once, so including in the same file as the plotting function might be better.

Speaking about more readable code, what makes my code bad in terms of readability? Do I use any bad practices? (I will ping @Islast as well) 
Appreciate your feedback, as it is really important and will allow me to improve the quality of the code.

**Comment by KirstieJane** [comment]
I don’t think you use any bad practices at all!! I think a _lot_ of pandas code is really hard to read! There are so many things that pandas can do that it ends up being hard to see what’s going on.

That’s why some documentation that says: here’s the information we have, in this shape, and here’s why we need it to be in a different shape (so seaborne interprets it properly) is probably the solution.

Have a google next week for pandas style guides....I haven’t seen one, and they might not exist, but that’s the only part that’s hard for me to wrap my head around. And it isn’t that your code is bad at all!

File: scona/visualisations.py [hunk 10]
```diff
@@ -0,0 +1,397 @@
...
+import numpy as np
+import networkx as nx
+import pandas as pd
+import matplotlib.pyplot as plt
+import seaborn as sns
+
+from scona.helpers import save_fig
+
+
+def plot_rich_club(brain_bundle, real_network, figure_name=None, color=None,
```
Comment (KirstieJane) [inline scona/visualisations.py]: Hi @wingedRuslan! I didn't have a huge amount of time over the weekend so I'm just doing a quick look through. One thing I've noticed is that there are some options that should be passed to this function as arguments to control the plotting. Eg for the legend and for the confidence intervals.

Can we make sure that its easy to control the legend text in the same way across functions? So however you'd adjust the legend text for the plot_network_measures function that should be the same for plot_rich_club.

See also #139 #138 

Thank you!

File: scona/visualisations.py [hunk 11]
```diff
@@ -0,0 +1,347 @@
...
+    if figure_name:
+        # use the helper-function from module helpers to save the figure
+        save_fig(fig, figure_name)
+        # close the file after saving to a file
+        plt.close(fig)
+
+
+def plot_network_measures(brain_bundle, real_network, figure_name=None,
+                          color=None, ci=95, show_legend=True):
+    """
```
Comment (Islast) [inline scona/visualisations.py]: I'm going to ask you to replace `real_network` with something else throughout. 
My problem with "real" is that it's a fairly dubious way to describe a heavily simplified mathematical model. I think it's being used here in place of a much longer-to-state notion such as _"the proper network, the one you got from the mri data_" or _"the network you were interested in in the first place, before you made all those random ones"_

 I've got a couple of suggestions, I don't love any of them, but I prefer them all to _real_
* `main_network`  or simply `network`- this doesn't say very much but that's okay
* `distinguished_network`, `emphasized_network`, `featured_network` or `highlighted_network` - these suggest what the plotting function is going to do with them, which is highlight them over the other networks
* `initial_network` or `original_network` - this tells us we're talking about the network from which all the other networks were created.

File: scona/visualisations.py (hunk 10 above)
Comment (Islast) [reply scona/visualisations.py]: #138 is resolved, although it has spun into a new complaint of my own about terminology

File: scona/visualisations.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Islast) [reply scona/visualisations.py]: I see now that it was I who introduced the term __real network__ the greatest monster is always in the mirror :moyai:

File: scona/visualisations.py (hunk 11 above)
Comment (wingedRuslan) [reply scona/visualisations.py]: I like your suggestion to rename "real_network" to "original_network"

File: scona/graph_measures.py [hunk 12]
```diff
@@ -346,15 +346,12 @@ def calculate_nodal_measures(
 
     # ==== CALCULATE MEASURES ====================
 
-    def calc_nodal_measure(G, measure, method, force=False):
```
[the remark this answers is not in the document]
Comment (wingedRuslan) [reply scona/graph_measures.py]: As for me, it just looked really strange to have a function `calc_nodal_measure` inside a bigger one - `calculate_nodal_measures` and calling this function `calc_nodal_measure` only once a few lines below.

@Islast If it is made intentionally, I can return the original code, no problem. Undo the changes and have the original code?

[the remark this answers is not in the document]
Comment (Islast) [reply scona/graph_measures.py]: I keep going back and forth on this, so I'm gonna defer to your second set of eyes. If you think it's more readable this way stick with it :+1:

File: scona/visualisations.py [hunk 13]
```diff
@@ -0,0 +1,352 @@
...
+        a python dictionary with BrainNetwork objects as values
+        (:class:`str`: :class:`BrainNetwork` pairs), contains original Graph
+        and random graphs.
+    original_network: str, required
+        This should index the particular network in `brain_bundle` that you want
+        the figure to highlight. A distribution of all the other networks in
+        `brain_bundle` will be rendered for comparison.
+        To plot rich club values along with the rich club values from
+        randomed graphs it is required to pass the name of the initial network,
+        the proper network, the one you got from the mri data.
```
Comment (Islast) [inline scona/visualisations.py]: I would like to remove the following from all function descriptions:
```
       To plot rich club values along with the rich club values from
        randomed graphs it is required to pass the name of the initial network,
        the proper network, the one you got from the mri data.
```

Comment (wingedRuslan) [reply scona/visualisations.py]: Removed

File: scona/visualisations.py (hunk 11 above)
Comment (KirstieJane) [reply scona/visualisations.py]: I've put in some changes in https://github.com/wingedRuslan/scona/pull/3 to catch this in a few other places.

**Comment by KirstieJane** [comment]
Hey @wingedRuslan! This all looks great to me! If you can merge https://github.com/wingedRuslan/scona/pull/6 then I'm totally fine for you to merge!

@Islast - you requested changes and I think they're all addressed. Would you mind taking a look and approving when you have a chance?

**Comment by wingedRuslan** [comment]
@KirstieJane, Thank you a lot!  I would like to merge but as @Islast requested changes, her approval is needed.

**Comment by KirstieJane** [comment]
Wooooo - click the big green button @wingedRuslan!! MERGE AWAY 🚀:star2:
